src/fx/TimeStretchPitchScaleReader.cpp

```diff
@@ -0,0 +1,285 @@
+/*******************************************************************************
+ * Copyright 2009-2025 Jörg Müller
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+#include "fx/TimeStretchPitchScaleReader.h"
+
+#include <iostream>
+
+#include "IReader.h"
+
+#include "util/Buffer.h"
+
+using namespace RubberBand;
+
+AUD_NAMESPACE_BEGIN
+
+TimeStretchPitchScaleReader::TimeStretchPitchScaleReader(std::shared_ptr<IReader> reader, double timeRatio, double pitchScale, StretcherQualityOptions quality) :
+    EffectReader(reader),
+    m_timeRatio(timeRatio),
+    m_pitchScale(pitchScale),
+    m_position(0),
+    m_length(0),
+    m_quality(-1),
+    m_finishedReader(false),
+    m_input(reader->getSpecs().channels),
+    m_processData(reader->getSpecs().channels),
+    m_output(reader->getSpecs().channels),
+    m_retrieveData(reader->getSpecs().channels)
+{
+	configure(quality);
+}
+
+void TimeStretchPitchScaleReader::read(int& length, bool& eos, sample_t* buffer)
+{
+	if(length == 0)
+		return;
+
+	if(m_pitchScale == 1.0 && m_timeRatio == 1.0)
+	{
+		m_reader->read(length, eos, buffer);
+		return;
+	}
```

Comment:
This will not work properly if the engine has been activated, since you might have fed it samples that are then skipped if the animation changes the values here to 1. But see my overall suggestion of the review which - if implemented - would make this code not necessary anymore as everything could be run through the engine if the effect is applied.